### PR TITLE
[COR-67] Add registry of services to allow better discovery and metadata

### PR DESCRIPTION
**Pull request by midhilkari**
Relates to: [COR-67] 

- Add Registry
- Version bump

[COR-67]: https://loke.atlassian.net/browse/COR-67

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -64,76 +64,124 @@ export interface Service {
...
+      service: this.serviceDetails.service,
+      help: this.serviceDetails.help,
+      path: `/rpc/${this.serviceDetails.service}`,
+    };
+  }
+}
 
-    requestCount.inc(requestMeta);
+export class Registry {
+  private registeredServices: Record<string, RegisteredService<any>> = {};
```
Comment (denwilliams) [inline index.ts]: note: need to figure out how to not have `any` here before anything final

File: test/express.ts [hunk 2]
```diff
@@ -117,3 +122,44 @@ Can include **Markdown**.`,
...
+  const meta: ServiceDetails<typeof service> = {
+    expose: [{ methodName: "hello" }],
+    service: "hello-service",
+    help: "hello",
+  };
+
+  registry.register(service, meta);
+
+  app.use(
+    "/.well-known/loke-rpc",
```
Comment (denwilliams) [inline test/express.ts]: Some "light" reading for standards on this

https://datatracker.ietf.org/doc/html/rfc5785

Comment (denwilliams) [reply test/express.ts]: https://www.iana.org/assignments/well-known-uris/well-known-uris.xhtml

Comment (denwilliams) [reply test/express.ts]: We could have a tiered approach, eg this comment

> The suffix "posh" is expected to be followed by an additional path component consisting of a service name (say, "spice") and a file extension of ".json", resulting in a full path of, for instance, "/.well-known/posh/spice.json". Registration of service names shall be requested by developers of the relevant application protocols.

eg `/.well-known/loke/rpc-server` `/.well-known/loke/rpc-client`

or  `/.well-known/loke-rpc/server` `/.well-known/loke-rpc/client`

Comment (midhilkari) [reply test/express.ts]: Think `/.well-known/loke-rpc/server`  `/.well-known/loke-rpc/client` is good

File: index.ts [hunk 3]
```diff
@@ -64,76 +64,126 @@ export interface Service {
...
+          failureCount.inc(Object.assign({ type: err.type }, requestMeta));
+          next(new RpcError(serviceName, methodName, err));
+        })
+        .finally(end);
+    };
+  }
 
-    const requestMeta = { handler: `${serviceName}.${methodName}` };
-    const end = requestDuration.startTimer(requestMeta);
+  toWellKnown() {
```
Comment (denwilliams) [inline index.ts]: Thinking maybe `toWellKnownMeta`

File: index.ts [hunk 4]
```diff
@@ -64,76 +64,126 @@ export interface Service {
...
+    return {
+      service: this.serviceDetails.service,
+      help: this.serviceDetails.help,
+      path: `/rpc/${this.serviceDetails.service}`,
```
Comment (denwilliams) [inline index.ts]: I think we want a way to override the path where some existing services may not follow this convention, eg
```suggestion
      path: this.serviceDetails.path || `/rpc/${this.serviceDetails.service}`,
```

File: index.ts [hunk 5]
```diff
@@ -64,76 +64,126 @@ export interface Service {
...
+        })
+        .finally(end);
+    };
+  }
 
-    const requestMeta = { handler: `${serviceName}.${methodName}` };
-    const end = requestDuration.startTimer(requestMeta);
+  toWellKnown() {
+    return {
+      service: this.serviceDetails.service,
```
Comment (denwilliams) [inline index.ts]: I reckon `name:` for this field

File: index.ts [hunk 6]
```diff
@@ -64,76 +64,126 @@ export interface Service {
...
+
+  createWellKnownHandler(): RequestHandler {
+    return (
+      req: { path: string; method: string; body: unknown },
+      res: { json: (body: unknown) => void }
+    ) => {
+      const requestMeta = { handler: `well-known-services` };
+      const end = requestDuration.startTimer(requestMeta);
+
+      requestCount.inc(requestMeta);
```
Comment (denwilliams) [inline index.ts]: Was undecided before, but now think that the metrics aren't very useful here.

I reckon get rid of them.

File: index.ts [hunk 7]
```diff
@@ -64,76 +64,126 @@ export interface Service {
...
+        ),
+      });
+
+      end();
+    };
+  }
 }
 
+export const registry = new Registry();
+export const wellKnownHandlerPath = "/.well-known/loke-rpc/server";
```
Comment (denwilliams) [inline index.ts]: String and number constants should really be UPPER_SNAKE_CASE.

And handler is the bit that handles the request, metadata is what the handler serves and the path represents.

```suggestion
export const WELL_KNOWN_META_PATH = "/.well-known/loke-rpc/server";
```

File: package.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@loke/http-rpc",
-  "version": "4.0.1",
+  "version": "5.0.0-beta",
```
Comment (denwilliams) [inline package.json]: TODO: once we are happy remove the beta

File: test/express.ts [hunk 9]
```diff
@@ -117,3 +122,46 @@ Can include **Markdown**.`,
...
+  const meta: ServiceDetails<typeof service> = {
+    expose: [{ methodName: "hello" }],
+    service: "hello-service",
+    help: "hello",
+  };
+
+  registry.register(service, meta);
+
+  app.use(
+    "/.well-known/loke-rpc/server",
```
Comment (denwilliams) [inline test/express.ts]: May as well show best practices here, use the const `WELL_KNOWN_META_PATH`

**Review by denwilliams — CHANGES_REQUESTED** [review]
I think you actually have most of the functionality we need here.

Some comments to address, and the README updated to reflect the changes.

Be clear about breaking changes between v4 and v5.

File: index.ts (hunk 3 above)
Comment (midhilkari) [reply index.ts]: yeah actually makes sense as we would be passing meta data everywhere this is used. Changing it now

File: index.ts (hunk 4 above)
Comment (midhilkari) [reply index.ts]: will have path as optional field in `serviceDetails` as we would not be explicitly passing path. if different then can pass path where this is used

File: index.ts (hunk 5 above)
Comment (midhilkari) [reply index.ts]: done

File: index.ts (hunk 6 above)
Comment (midhilkari) [reply index.ts]: done

File: index.ts (hunk 7 above)
Comment (midhilkari) [reply index.ts]: done

File: test/express.ts (hunk 9 above)
Comment (midhilkari) [reply test/express.ts]: fixing this now

File: README.md [hunk 10]
```diff
@@ -18,26 +22,31 @@ const MY_SERVICE_META = {
...
-  myService,
-  MY_SERVICE_META
-);
+//BREAKING CHANGES:
+// "createRequestHandler" was renamed to createWellKnownHandler and can be accesed through register()
+// Added new createWellKnownHandler which serves the metadata
+// WELL_KNOWN_META_PATH represents the path where meta data served
+
+const myRpcService = registry.register(myService, MY_SERVICE_META)
+  .createWellKnownHandler;
```
Comment (denwilliams) [inline README.md]:
```suggestion
  .createWellKnownHandler();
```

Also I think this should be `createWellKnownMetaHandler`

Comment (denwilliams) [reply README.md]: Actually this is the service, isn't this `createRequestHandler`?

Comment (midhilkari) [reply README.md]: Oops. mixed up things here!!

File: README.md [hunk 11]
```diff
@@ -18,26 +22,31 @@ const MY_SERVICE_META = {
...
+
+const wellKnownHandler = registry.createWellKnownHandler();
 
-const errorLogger = msg => console.log(msg);
+const errorLogger = (msg) => console.log(msg);
 
 app.use("/rpc", myRpcService);
-app.use(lokeHttpRpc.createErrorHandler({ log: errorLogger }));
+app.get(WELL_KNOWN_META_PATH, wellKnownHandler);
```
Comment (denwilliams) [inline README.md]: IMO this is less clear than just going 
```suggestion
app.get(WELL_KNOWN_META_PATH, registry.createWellKnownHandler());
```

File: README.md [hunk 12]
```diff
@@ -18,26 +22,31 @@ const MY_SERVICE_META = {
...
+    },
+  ],
 };
 
-const myRpcService = lokeHttpRpc.createRequestHandler(
-  myService,
-  MY_SERVICE_META
-);
+//BREAKING CHANGES:
+// "createRequestHandler" was renamed to createWellKnownHandler and can be accesed through register()
```
Comment (denwilliams) [inline README.md]: No, this is wrong. It is still createRequestHandler but you need to register the service first.

File: README.md [hunk 13]
```diff
@@ -18,26 +22,31 @@ const MY_SERVICE_META = {
...
+// Added new createWellKnownHandler which serves the metadata
+// WELL_KNOWN_META_PATH represents the path where meta data served
```
Comment (denwilliams) [inline README.md]: Don't do this as comments. Put this right at the top as a heading, eg

```
## Breaking Changes for v5

Services need to be registered before a request handler can be created.

### Migrating from v4 to v5

... add before/after
```

File: README.md [hunk 14]
```diff
@@ -18,26 +22,31 @@ const MY_SERVICE_META = {
...
+  ],
 };
 
-const myRpcService = lokeHttpRpc.createRequestHandler(
-  myService,
-  MY_SERVICE_META
-);
+//BREAKING CHANGES:
+// "createRequestHandler" was renamed to createWellKnownHandler and can be accesed through register()
+// Added new createWellKnownHandler which serves the metadata
```
Comment (denwilliams) [inline README.md]: createWellKnownHandler is not a breaking change, it is an addition.

It just needs to be added to the documentation too.

File: README.md (hunk 11 above)
Comment (midhilkari) [reply README.md]: yeah this looks much clear. Updating this

File: index.ts [hunk 15]
```diff
@@ -58,82 +58,126 @@ export interface ServiceDetails<S> {
...
-    multiArg: false,
-    help: serviceHelp,
-    interfaces: exposed.map((method: MethodDetails<S>) => {
-      if (typeof method === "string") {
-        throw new Error(
-          "Schema for expose has changed. Please refer to @loke/http-rpc documentation."
+class RegisteredService<S extends Service> {
+  constructor(private service: S, private serviceDetails: ServiceDetails<S>) {}
+
+  createRequestHandler(): RequestHandler {
```
Comment (denwilliams) [inline index.ts]: This is a library so we should be adding method descriptions to help devs use it

```suggestion
  /**
   * Creates an express HTTP handler that will process RPC requests for methods exposed by the service, as well as return metadata about the service.
   */
  createRequestHandler(): RequestHandler {
```

File: index.ts [hunk 16]
```diff
@@ -58,82 +58,126 @@ export interface ServiceDetails<S> {
...
+        .then((result) => (console.log("result", result), res.json(result)))
+        .catch((err) => {
+          failureCount.inc(Object.assign({ type: err.type }, requestMeta));
+          next(new RpcError(serviceName, methodName, err));
+        })
+        .finally(end);
+    };
+  }
+
+  toWellKnownMeta() {
```
Comment (denwilliams) [inline index.ts]: Again, doc comments for anything that is public

```suggestion
  /**
   * ... add description here
   */
  toWellKnownMeta() {
```

File: index.ts [hunk 17]
```diff
@@ -58,82 +58,126 @@ export interface ServiceDetails<S> {
...
+}
 
-    const requestMeta = { handler: `${serviceName}.${methodName}` };
-    const end = requestDuration.startTimer(requestMeta);
+export class Registry {
+  // TODO: Fix the RegisteredService type from 'any'
+  private registeredServices: Record<string, RegisteredService<any>> = {};
 
-    requestCount.inc(requestMeta);
+  register<S extends Service>(
```
Comment (denwilliams) [inline index.ts]: doc comment here

File: index.ts [hunk 18]
```diff
@@ -58,82 +58,126 @@ export interface ServiceDetails<S> {
...
-        next(new RpcError(serviceName, methodName, err));
-      })
-      .finally(end);
-  };
+    this.registeredServices[serviceDetails.service] = rs;
+
+    return rs;
+  }
+
+  createWellKnownHandler(): RequestHandler {
```
Comment (denwilliams) [inline index.ts]:
```suggestion
  createWellKnownMetaHandler(): RequestHandler {
```

and doc comment

File: index.ts [hunk 19]
```diff
@@ -58,82 +58,126 @@ export interface ServiceDetails<S> {
...
+      res.json({
+        services: Object.values(this.registeredServices).map((rs) =>
+          rs.toWellKnownMeta()
+        ),
+      });
+    };
+  }
 }
 
+export const registry = new Registry();
```
Comment (denwilliams) [inline index.ts]:
```suggestion
/** Default registry */
export const registry = new Registry();
```

File: index.ts [hunk 20]
```diff
@@ -58,82 +58,126 @@ export interface ServiceDetails<S> {
...
+export const WELL_KNOWN_META_PATH = "/.well-known/loke-rpc/server";
```
Comment (denwilliams) [inline index.ts]:
```suggestion
/** Default path to expose discovery metadata on a well-known URL *.
export const WELL_KNOWN_META_PATH = "/.well-known/loke-rpc/server";
```

Comment (midhilkari) [reply index.ts]: done

File: index.ts (hunk 19 above)
Comment (midhilkari) [reply index.ts]: adding comments now

File: index.ts (hunk 18 above)
Comment (midhilkari) [reply index.ts]: Adding comments now

File: index.ts (hunk 17 above)
Comment (midhilkari) [reply index.ts]: Adding comments now

File: index.ts (hunk 16 above)
Comment (midhilkari) [reply index.ts]: Adding comments now

File: index.ts (hunk 15 above)
Comment (midhilkari) [reply index.ts]: Adding comments now

File: README.md (hunk 14 above)
Comment (midhilkari) [reply README.md]: Fixing this now

File: README.md (hunk 13 above)
Comment (midhilkari) [reply README.md]: done

File: README.md (hunk 12 above)
Comment (midhilkari) [reply README.md]: Fixed this

File: README.md [hunk 21]
```diff
@@ -1,15 +1,61 @@
...
+const {
+  registry,
+  createErrorHandler,
+  WELL_KNOWN_META_PATH,
+} = require("loke-http-rpc");
+
+const myRpcService = registry.register(myService, MY_SERVICE_META)
+  .createRequestHandler;
+
+app.use("/rpc", myRpcService);
```
Comment (denwilliams) [inline README.md]:
```suggestion
app.use("/rpc/service-name", myRpcService);
```

File: README.md [hunk 22]
```diff
@@ -1,15 +1,72 @@
...
 
+## Breaking Changes for v5
+
+Services need to be registered before a request handler can be created.
+
+### Migrating from v4 to v5
+
+- Services now needs to be registered before creating a request handler
+- Added new well know handler ("createWellKnownMetaHandler()") that serves service metadata
+- Exposing well-known URL for uniformity across the system to access service metadata
```
Comment (denwilliams) [inline README.md]: Also

- You can now create a request handler for multiple services contained in a separate registry

File: package.json [hunk 23]
```diff
@@ -38,13 +42,11 @@
     "got": "^11.6.2",
     "prettier": "2.1.1",
     "prom-client": "^12.0.0",
+    "router": "^1.3.5",
```
Comment (denwilliams) [inline package.json]: This is not a dev dependency - it is used for the actual library code, not just tests

**Review by denwilliams — COMMENTED** [review]
Looking pretty close now. I think we can take the `-beta` off soon and get it out.

Couple of comments below.

File: README.md (hunk 22 above)
Comment (midhilkari) [reply README.md]: added this now

File: package.json (hunk 23 above)
Comment (midhilkari) [reply package.json]: fixed this. added back to dependency list.

**Review by denwilliams — APPROVED** [review]

File: package-lock.json [hunk 24]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@loke/http-rpc",
-  "version": "4.0.1",
+  "version": "5.0.0-beta",
```
Comment (denwilliams) [inline package-lock.json]:
```suggestion
  "version": "5.0.0",
```

Comment (midhilkari) [reply package-lock.json]: fixed this

**Review by denwilliams — APPROVED** [review]

**Pull request merged**